### DISCUSSION TITLE
Potential bug with CreateIndexes(Async) when updating indexes

### DISCUSSION THREAD

**Discussion by danielwinkler** [ravendb/ravendb#19504] [Q&A]
We are using the following code to deploy / update indexes

```csharp
private async Task UpdateIndexes()
{
    var indexAssembly = typeof(MyIndex_SearchIndex).Assembly;
    await IndexCreation.CreateIndexesAsync(indexAssembly, _store);
    _logger.LogInformation($"Indexes created successfully");
}
```

I noticed that the index is sometimes not updated; some examples:

- I add a `Index(x => x.SomeField, FieldIndexing.Exact);` directive to the C# index definition
- I remove part of the index definition directly in the studio, so that they are out of sync with the C# definition

To verify that the call is actually being performed:

- There is no issue when i delete the index server side, it gets correctly deployed
- If i change the index server side to throw an error on overwrite, it throws an error; if i remove the option it doesn't throw, but also doesn't update

RavenDB.Client Version="6.2.0"
Server version 6.0.106

**Comment by arekpalinski** [comment]
Hi,

Does it happen consistently? I assume it doesn't because you said:

> sometimes not updated

Are you able to provide reproduction steps which result in this behavior? Can you think about any additional circumstances that might cause this? Is there any chance that you might "sometimes" have Locked Ignored option set?

**Reply by danielwinkler** [reply]
I did not investigate when it happens, that's why i wrote "sometimes"
I just rolled out a new index, and it also did not update.

I checked the logs and it does trigger the update but might not actually replace it properly.

- Log of replacement attempt: It recognizes that there is a change but the index on database does not update
[admin-log-2024-10-28 08-12.txt](https://github.com/user-attachments/files/17538921/admin-log-2024-10-28.08-12.txt)

- Log of deployment after manually deleting the index
[admin-log-2024-10-28 08-14.txt](https://github.com/user-attachments/files/17538925/admin-log-2024-10-28.08-14.txt)

The index is very simple:

```csharp
public class Voice_AllByNameIndex : AbstractIndexCreationTask<Voice>
{
    public record Entry
    {
        public required string Name { get; init; }
        public required CloudProvider Provider { get; init; }
    }

    public Voice_AllByNameIndex()
    {
        Map = voices => from voice in voices
                        select new Entry
                        {
                            Name = voice.Name,
                            Provider = voice.Provider
                        };
    }
}
```

The change was adding the Provider.

**Reply by arekpalinski** [reply]
Indeed it looks that the index got updated:

```
2024-10-28T07:12:54.8310668Z, 19, Information, A, Raven.Server.Rachis.RachisConsensus, Apply of PutIndexCommand with index 560181428 was successful. AdditionalDebugInformation: Index name: 'Voice/AllByNameIndex' for database 'DATABASE_NAME'.
```

Can you please check the index history?

![image](https://github.com/user-attachments/assets/8e8e7a3d-60c6-4702-acf7-d4b893866858)

**Reply by danielwinkler** [reply]
I was waiting to reproduce it again because i couldn't after starting the ticket 

However, in the end it was my own mistake, as i deleted all indexes that were not needed any more from the database ... unfortunately this logic also covered the "ReplacementOf/" indexes ... so they were there but depending on how long the indexing took , could be getting deleted before replacing the old definition
